sockeye/data_io.py

```diff
@@ -495,211 +1240,166 @@ def __init__(self,
-            self.bucket_batch_sizes[buck_idx] = BucketBatchSize(batch_size_seq, batch_size_word)
-            # Track largest batch size by total elements
-            largest_total_batch_size = max(largest_total_batch_size, batch_size_seq * max(*bucket_shape))
-        # Final step: guarantee that largest bucket by sequence length also has largest total batch size.
-        # When batching by sentences, this will already be the case.
-        if self.batch_by_words:
-            padded_seq_len = max(*self.buckets[-1])
-            average_seq_len = self.data_target_average_len[-1]
-            while self.bucket_batch_sizes[-1].batch_size * padded_seq_len < largest_total_batch_size:
-                self.bucket_batch_sizes[-1] = BucketBatchSize(
-                    self.bucket_batch_sizes[-1].batch_size + self.batch_num_devices,
-                    self.bucket_batch_sizes[-1].average_words_per_batch + self.batch_num_devices * average_seq_len)
-
-    def _convert_to_array(self):
-        for i in range(len(self.data_source)):
-            self.data_source[i] = np.asarray(self.data_source[i], dtype=self.dtype)
-            self.data_target[i] = np.asarray(self.data_target[i], dtype=self.dtype)
-            self.data_label[i] = np.asarray(self.data_label[i], dtype=self.dtype)
-
-            n = len(self.data_source[i])
-            batch_size_seq = self.bucket_batch_sizes[i].batch_size
-            if n % batch_size_seq != 0:
-                buck_shape = self.buckets[i]
-                rest = batch_size_seq - n % batch_size_seq
-                if self.fill_up == 'pad':
-                    raise NotImplementedError
-                elif self.fill_up == 'replicate':
-                    logger.info("Replicating %d random sentences from bucket %s to size it to multiple of %d", rest,
-                                buck_shape, batch_size_seq)
-                    random_indices = np.random.randint(self.data_source[i].shape[0], size=rest)
-                    self.data_source[i] = np.concatenate((self.data_source[i], self.data_source[i][random_indices, :]),
-                                                         axis=0)
-                    self.data_target[i] = np.concatenate((self.data_target[i], self.data_target[i][random_indices, :]),
-                                                         axis=0)
-                    self.data_label[i] = np.concatenate((self.data_label[i], self.data_label[i][random_indices, :]),
-                                                         axis=0)
+                current_shard_fname = self.shards_fnames[self.shard_index]
+            remaining_shards = [shard for shard in self.shards_fnames if shard != current_shard_fname]
+            next_shard_fname = random.choice(remaining_shards)
+            remaining_shards = [shard for shard in self.shards_fnames if shard != next_shard_fname]
+            random.shuffle(remaining_shards)
+
+            self.shards_fnames = [next_shard_fname] + remaining_shards
```

Comment:
I'm probably missing something here, but is this (and the block before) not equivalent to self.shard_fnames = random.shuffle(self.shards_fnames)?

Reply:
the idea is to prevent the case where the last shard in an epoch is picked as the first shard in the next epoch.

Reply:
Oh, I see. A comment would help ;-)

Reply:
Line 1303 :)